runtime/memprof.c

```diff
@@ -125,259 +130,393 @@ static uintnat mt_generate_binom(uintnat len)
   return res;
 }
 
-/**** Interface with the OCaml code. ****/
+/**** Capturing the call stack *****/
 
-static void purge_postponed_queue(void);
+/* This function is called for postponed blocks, so it guarantees
+   that the GC is not called. */
+static value capture_callstack_postponed(void)
+{
+  value res;
+  uintnat wosize = caml_current_callstack_size(callstack_size);
+  /* We do not use [caml_alloc] to make sure the GC will not get called. */
+  if (wosize == 0) return Atom(0);
+  res = caml_alloc_shr_no_track_noexc(wosize, 0);
+  if (res != 0) caml_current_callstack_write(res);
+  return res;
+}
 
-CAMLprim value caml_memprof_set(value v)
+/* Should be called with [caml_memprof_suspended == 1] */
+static value capture_callstack(void)
 {
-  CAMLparam1(v);
-  double l = Double_val(Field(v, 0));
-  intnat sz = Long_val(Field(v, 1));
+  value res;
+  uintnat wosize = caml_current_callstack_size(callstack_size);
+  CAMLassert(caml_memprof_suspended);
+  res = caml_alloc(wosize, 0);
+  caml_current_callstack_write(res);
+  return res;
+}
 
-  if (sz < 0 || !(l >= 0.) || l > 1.) /* Checks that [l] is not NAN. */
-    caml_invalid_argument("caml_memprof_set");
+/**** Data structures for tracked blocks. ****/
 
-  /* This call to [caml_memprof_set] may stop sampling or change the
-     callback. We have to make sure that the postponed queue is empty
-     before continuing. */
-  if (!caml_memprof_suspended)
-    caml_raise_if_exception(caml_memprof_handle_postponed_exn());
-  else
-    /* But if we are currently running a callback, there is nothing
-       else we can do than purging the queue. */
-    purge_postponed_queue();
+/* During the alloc callback for a minor allocation, the block being
+   sampled is not yet allocated. Instead, it's represented as this. */
+#define Placeholder_value (Val_long(0x42424242))
 
-  if (!init) {
-    int i;
-    init = 1;
+/* When an entry is deleted, its index is replaced by that integer. */
+#define Invalid_index (~(uintnat)0)
 
-    mt_index = 624;
-    mt_state[0] = 42;
-    for(i = 1; i < 624; i++)
-      mt_state[i] = 0x6c078965 * (mt_state[i-1] ^ (mt_state[i-1] >> 30)) + i;
+struct tracked {
+  /* Memory block being sampled. */
+  value block;
 
-    caml_register_generational_global_root(&memprof_callback);
-  }
+  /* Number of samples in this block. */
+  uintnat occurrences;
 
-  lambda = l;
-  if (l > 0) {
-    one_log1m_lambda = l == 1 ? 0 : 1/caml_log1p(-l);
-    next_mt_generate_binom = mt_generate_geom();
-  }
+  /* The header of this block (useful for tag and size) */
+  header_t header;
 
-  caml_memprof_renew_minor_sample();
+  /* The value returned by the previous callback for this block, or
+     the callstack if the alloc callback has not been called yet. */
+  value user_data;
 
-  callstack_size = sz;
+  /* Pointer to the [t_idx] variable in the [run_callback] frame
+     which is currently running the callback for this entry. */
```

Comment:
> there are situations where reentrancy state is saved before calling a callback.

This is not the case yet, but I anticipate that this will be the case when sampling in native mode will be implemented.